File: backend/app/schemas/execution.py
"""
Pydantic schemas for execution (runs).
"""
from datetime import datetime
from typing import Optional, Any
from uuid import UUID

from pydantic import BaseModel, Field


class RunCreate(BaseModel):
    trigger_type: str = Field(default="manual", pattern="^(manual|webhook|schedule)$")
    graph_version_id: Optional[UUID] = None


class RunResponse(BaseModel):
    id: UUID
    graph_id: UUID
    graph_version_id: Optional[UUID]
    trigger_type: str
    triggered_by: Optional[UUID]
    status: str
    error_message: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RunListResponse(BaseModel):
    runs: list[RunResponse]
    total: int


class RunNodeResponse(BaseModel):
    id: UUID
    run_id: UUID
    node_id: str
    node_type: str
    status: str
    attempt_count: int
    max_retries: int
    input_data: Optional[dict] = None
    output_data: Optional[dict] = None
    items_processed: Optional[int] = None
    duration_ms: Optional[int] = None
    error_message: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class RunDetailResponse(BaseModel):
    run: RunResponse
    run_nodes: list[RunNodeResponse] = Field(default_factory=list)


class RunControlRequest(BaseModel):
    """Request body for run control operations (cancel, pause, resume, retry)."""
    pass  # No additional parameters needed for Phase 1

File: frontend/src/lib/types/execution.ts
/**
 * Execution (run) types matching backend API.
 */

export type RunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
export type TriggerType = 'manual' | 'webhook' | 'schedule';
export type NodeRunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'skipped';

export interface Run {
  id: string;
  graph_id: string;
  graph_version_id: string | null;
  trigger_type: TriggerType;
  triggered_by: string | null;
  status: RunStatus;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface RunNode {
  id: string;
  run_id: string;
  node_id: string;
  node_type: string;
  status: NodeRunStatus;
  attempt_count: number;
  max_retries: number;
  input_data: Record<string, unknown> | null;
  output_data: Record<string, unknown> | null;
  items_processed: number | null;
  duration_ms: number | null;
  error_message: string | null;
  created_at: string;
}

export interface RunDetail {
  run: Run;
  run_nodes: RunNode[];
}

export interface RunListResponse {
  runs: Run[];
  total: number;
}

export interface RunCreate {
  trigger_type?: TriggerType;
  graph_version_id?: string;
}

// WebSocket event types
export type WsEventType =
  | 'run_started'
  | 'run_completed'
  | 'run_failed'
  | 'run_cancelled'
  | 'node_started'
  | 'node_completed'
  | 'node_failed'
  | 'node_progress'
  | 'node_retried';

export interface WsEvent {
  type: WsEventType;
  run_id: string;
  data: Record<string, unknown>;
}
